### PR TITLE
crimson/os/seastore: implement metrics at cache level

### PR DESCRIPTION
**Pull request by cyx1231st**
This is the follow-up work of https://github.com/ceph/ceph/pull/42202

It's to profile seastore at cache layer:
* cache hit ratio at the granularity of sub-component and categorized by transaction source;
* transactional read/mutate/fresh/retire efforts that are discarded and committed;
* transactional read effort that is successful;
* dirty and cached extent usage;

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/os/seastore/cache.cc [hunk 1]
```diff
@@ -245,6 +246,77 @@ void Cache::register_metrics()
...
+            ),
+          }
+        );
+      }
+    };
+
+  for (auto& [src, label] : labels_by_src) {
+    for (auto& [ext, _label] : labels_by_ext) {
+      if (ext != extent_types_t::RETIRED_PLACEHOLDER) {
+        register_cache_access(src, ext);
```
Comment (tchaikov) [inline src/crimson/os/seastore/cache.cc:311]: why not just pass `label` and `label_` (which deserves a better name, by the way) to  `register_cache_access()`?

File: src/crimson/os/seastore/cache.cc [hunk 2]
```diff
@@ -245,6 +246,77 @@ void Cache::register_metrics()
...
+  /*
+   * cache_query: cache_access and cache_hit
+   */
+  auto register_cache_access =
+    [this, &labels_by_src, &labels_by_ext](src_t src, extent_types_t ext) {
+      auto m_key = std::make_pair(src, ext);
+      stats.cache_query[m_key] = {0, 0};
+      std::ostringstream oss_labels;
+      oss_labels << "(src=" << src << ", ext=" << ext << ")";
+      auto str_labels = oss_labels.str();
```
Comment (tchaikov) [inline src/crimson/os/seastore/cache.cc:284]: i'd avoid encoding labels into the description. just to minimize the repetition of information.

File: src/crimson/os/seastore/cache.cc [hunk 3]
```diff
@@ -245,6 +246,77 @@ void Cache::register_metrics()
       ),
     }
   );
+
+  /*
+   * cache_query: cache_access and cache_hit
+   */
```
Comment (tchaikov) [inline src/crimson/os/seastore/cache.cc:277]: nit, 
```c++
// cache_query: cache_access and cache_hit
```

File: src/crimson/os/seastore/cache.h [hunk 4]
```diff
@@ -584,15 +591,63 @@ class Cache {
...
 
-  using src_ext_t = std::pair<Transaction::src_t, extent_types_t>;
+  struct query_counters_t {
+    uint64_t access;
+    uint64_t hit;
+  };
+
+  struct effort_t {
+    uint64_t extents = 0;
+    uint64_t bytes = 0;
```
Comment (tchaikov) [inline src/crimson/os/seastore/cache.h:601]: it'd be difficult to tell what `extents` and `bytes` is for in this context. the same applies to `effort_t`.

File: src/crimson/os/seastore/cached_extent.h [hunk 5]
```diff
@@ -541,17 +542,27 @@ class ExtentIndex {
       extent.get_length());
     ceph_assert(a == b);
 
-    extent_index.insert(extent);
+    auto [iter, inserted] = extent_index.insert(extent);
```
Comment (tchaikov) [inline src/crimson/os/seastore/cached_extent.h:545]:
```suggestion
    [[maybe_unused]] auto [iter, inserted] = extent_index.insert(extent);
```

File: src/crimson/os/seastore/cache.h (hunk 4 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/cache.h:601]: The effort is counted by the number of extents and the bytes of these extents. So would `num_extents` and `bytes_extents` better?

And is `trans_effort_t` a better name?

Comment (tchaikov) [reply src/crimson/os/seastore/cache.h:601]: i mean, probably we could add a comment explaining that this represents the total number of extents and the total bytes of them involved in transactions, they are categorized by different sets of extents involved by transactions. like read_set , retireset, ...

Comment (cyx1231st) [reply src/crimson/os/seastore/cache.h:601]: Oh, sure. I'll add them.

File: src/crimson/os/seastore/cache.cc (hunk 1 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/cache.cc]: fixed.

File: src/crimson/os/seastore/cache.cc (hunk 2 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/cache.cc]: All removed in this PR.

File: src/crimson/os/seastore/cache.cc (hunk 3 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/cache.cc:277]: I wanted to use /* ... */ to emphasize the section of metrics to be registered.

File: src/crimson/os/seastore/cache.h (hunk 4 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/cache.h:611]: done.

File: src/crimson/os/seastore/cached_extent.h (hunk 5 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/cached_extent.h]: done.

**Review by athanatos — APPROVED** [review]

**Review by tchaikov — APPROVED** [review]

**Pull request merged**